rhombus-lib/rhombus/private/amalgam/closeable.rhm

```diff
@@ -0,0 +1,40 @@
+#lang rhombus/private/amalgam/core
+
+import:
+  "core-meta.rkt" open
+  "closeable.rkt" open
+
+export:
+  Closeable
+
+namespace Closeable:
+  export:
+    close
+    rename cdef as def  
+
+  fun close(v :: Closeable):
+    v.close()
+
+  defn.sequence_macro 'cdef $(bind :: Identifier) = $(rhs :: expr_meta.Parsed)
+                       $body
+                       ...':
+    ~op_stx: self
+    def rhs_statinfo = statinfo_meta.gather(rhs)
+    def resource = Syntax.make_temp_id("resource")
+    def resource_with_statinfo = statinfo_meta.wrap(resource, rhs_statinfo)
+    values(
+      '
+        block:
+          def mutable $resource = #false
+          try:
+            ~initially:
+              $resource := ($rhs $('::'.relocate(self)) Closeable)
+            block:
+              let $bind = $resource_with_statinfo
+              #void
+              $body
+              ...
+            ~finally:
+              ($resource :~ Closeable).close()
```

Comment:
IIUC, with this implementation, control is allowed to escape from and reenter the `body`, and, if so, the `rhs` is evaluated again to re-allocate the resource. Is this the semantics you want? I think a more useful default might be to wrap body in a continuation barrier to forbid reentry. (This would be stricter than `call-with-{in,out}put-file*`, which let you reenter, you just find your port closed when you do.) Reentrancy can be very useful, but, if a `Closeable` implementer isn't thinking about it, unfortunate consequences seem likely: for example, truncating the output you wrote to the file before the suspend/resume.

Reply:
Allowing jumps out and in with reopening was the intent, at least. I'm not certain that's the best choice, but I'm having trouble thinking of a situation where continuation jumps might happen and this boundary is the right place to block them.

Reply:
I had always wondered why `call-with-input-file` and similar don't support continuation jumps in by reopening a file. Now that I try it, the reason seems clear: `call-with-input-file` could reopen a file, but it won't change the port that the argument to `call-with-input-file` is using. So, I concede that a barrier here is a good idea.